saleor/graphql/checkout/mutations.py

```diff
@@ -0,0 +1,218 @@
+        if shipping_address:
+            shipping_address.save()
+            instance.shipping_address = shipping_address
+        super().save(info, instance, cleaned_input)
+        instance.save(update_fields=['shipping_address'])
+        variants = cleaned_input.get('variants')
+        quantities = cleaned_input.get('quantities')
+        if variants and quantities:
+            for variant, quantity in zip(variants, quantities):
+                add_variant_to_cart(instance, variant, quantity)
+
+
+
+class CheckoutLinesAdd(BaseMutation):
+    class Arguments:
+        checkout_id = graphene.ID(description='Checkout ID', required=True)
+        lines = graphene.List(
+            CheckoutLineInput, description='Checkout lines', required=True)
+
+    checkout = graphene.Field(Checkout, description='An updated checkout')
+
+    class Meta:
+        description = 'Adds a checkout line to existing checkout'
+
+    @classmethod
+    def mutate(cls, root, info, checkout_id, lines, replace=False):
+        checkout = get_node(info, checkout_id, only_type=Checkout)
+        variants, quantities = None, None
+        errors = []
+        if lines:
+            variant_ids = [line.get('variant_id') for line in lines]
+            variants = cls.get_nodes_or_error(
+                ids=variant_ids, errors=errors, field='variant_id',
+                only_type=ProductVariant)
+            quantities = [line.get('quantity') for line in lines]
+            line_errors = check_lines_quantity(variants, quantities)
+            if line_errors:
+                for err in line_errors:
+                    cls.add_error(errors, field=err[0], message=err[1])
+
+        if variants and quantities:
+            for variant, quantity in zip(variants, quantities):
+                add_variant_to_cart(checkout, variant, quantity, replace=replace)
+        return CheckoutLinesAdd(checkout=checkout, errors=errors)
+
+
+class CheckoutLinesUpdate(CheckoutLinesAdd):
+    checkout = graphene.Field(Checkout, description='An updated checkout')
+
+    class Meta:
+        description = 'Updates checkout line in existing checkout'
+
+    @classmethod
+    def mutate(cls, root, info, checkout_id, lines):
+        return super().mutate(root, info, checkout_id, lines, replace=True)
+
+
+class CheckoutLineDelete(BaseMutation):
+    class Arguments:
+        checkout_id = graphene.ID(description='Checkout ID', required=True)
+        line_id = graphene.ID(description='Checkout line ID')
+
+    checkout = graphene.Field(Checkout, description='An updated checkout')
+
+    class Meta:
+        description = 'Deletes a checkout line'
+
+    @classmethod
+    def mutate(cls, root, info, checkout_id, line_id):
+        errors = []
+        checkout = cls.get_node_or_error(
+            info, checkout_id, errors, 'checkout_id', only_type=Checkout)
+        line = cls.get_node_or_error(
+            info, line_id, errors, 'line_id', only_type=CheckoutLine)
+        line.delete()
```

Comment:
Two checks should be added before we can delete the line:
- ensure that this line belongs to this checkout
- ensure that the `line` is not None- it could be if there were errors returned by `get_node_or_error`.

Ideally we would pass a queryset of `checkout.lines.all()` to `get_nodes_or_error()` just as we do with `get_object_or_404()`, but that's a general improvement I want to add later.